### ISSUE TITLE
Issue when using AddApiExplorer and JsonTranscoding

### ISSUE THREAD

**Issue by mcastagna-sa** [dotnet/aspnet-api-versioning#1109]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

Hey guys, i'm having a problem with a project i'm working on and i wanted to see if i'm missing anything or maybe i've stumbled on a bug. 

I have created a minimal project where i'm able to reproduce the behaviour, I'm goint to try to guide you pasting snippets of code here.

When i run the program, and try to hit the endpoint v2 with postman everything goes as expected.

![image](https://github.com/user-attachments/assets/ea2fccb8-18e1-4943-9c65-155b44f5c89e)

But, instead, if i call v1, which should be responded by the jsontranscode, i get 404.

![image](https://github.com/user-attachments/assets/06491010-d5d9-462f-8030-1e64786b4279)

I've tracked this down to the call of AddApiExplorer. If i remove that call i'm able to get the v1 as i should. 

What do you think about this? Is this the expected behaviour? Am i doing something wrong? Is this a bug? If it is, is a bug in the jsontranscoding or asp.versioning library? 

Thank you so much for your time and sorry the lenght of this post. Let me know if you need anything else that could help you answering this.

### Expected Behavior

Endpoint is found and executed.

### Steps To Reproduce

Program.cs 
```
static void Main(string[] args)
 {
     var app = CreateHostBuilder(args).Build();

     app.UseRouting();

     app.MapControllers();

     app.MapGrpcService<GreeterService>();

     app.Run();
 }
  
 private static WebApplicationBuilder CreateHostBuilder(string[] args)
 {
     var builder = WebApplication.CreateBuilder(args);

     builder
         .WebHost.ConfigureKestrel((context, options) =>
         {
             options.ListenLocalhost(8080);
         });

     builder.Services.AddControllers();

     builder.Services.AddApiVersioning(options =>
          {
              options.ReportApiVersions = true;
          })
          .AddApiExplorer(options =>
          {
              options.GroupNameFormat = "'v'VVV";
              options.SubstituteApiVersionInUrl = true;                     
          });

     builder.Services
         .AddGrpc()
         .AddJsonTranscoding();

     return builder;
 }
````

V2Controller.cs

```
[ApiController]
[ApiVersion("2.0")]
[Produces("application/json")]
[Route("api/v{version:apiVersion}/controller/{param}")]
public class V2Controller : Controller
{
    [HttpGet]
    [Route("route")]
    public async Task<IActionResult> GetEndpoint()
    {
        return Ok("ok get v2");
    }
}
```
And greet.proto (This is the wiki example as you can see, just added the route i need to demostrate my point)

```
syntax = "proto3";

option csharp_namespace = "TestVersioningGRPC.Proto";
import "google/api/annotations.proto";

package greet;

service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {        
      get: "/api/v1/controller/{name}/route"
    };
  }
}

// The request message containing the user's name.
message HelloRequest {
  string name = 1;
}

message HelloReply {
  string message = 1;
}
```

### .NET Version

8.0.304

**Comment by commonsensesoftware** [comment]
I would be very surprised that anything about `AddApiExplorer` caused this behavior. It doesn't change any routing behavior whatsoever. It's purely discovery. I can't say I have much firsthand experience with JSON transcoding with gRPC, but there are some incongruencies I see that don't make a lot of sense.

1. Why use `Controller` (usually full MVC with views) instead of `ControllerBase`
2. Why are there two different routes to `GetEndpoint`?
3. Why does V2 have a route parameters with no parameter? 
    a. I don't even see how this can work since the parameter is required

I would have expected `2.0` to look something like:

```c#
[ApiController]
[ApiVersion(2.0)]
[Produces("application/json")]
[Route("api/v{version:apiVersion}/controller")]
public class V2Controller : Controller
{
    [HttpGet("{param}/route""]
    public IActionResult GetEndpoint(string param) => Ok("ok get v2");
}
```

>You can include `ApiVersion` as parameter if you don't want to hardcode `v2`

I'm not sure I understand exactly how ProtoBuf matches up to a controller. Is that code-generated?

I suspect that the next problem is a mismatch in names. API Version collation is performed on _logic name_, not route template. Route template may seem intuitive, but there are many nuances. `order/{id}` and `order/{id:int}` are not identical, but are semantically equivalent. Should `order/{id}/items` be part of the _Orders_ or _LIne Items_ (if there is one) API?

By default, ASP.NET Core uses the suffix convention `Controller`. If this present in the type name, it is removed. A type named `Controller` will end up as `""` and a type named `V2Controller` will end up as `"V2"`. Unless you otherwise configure it (and there are a few ways), this will be the logical name of the controller and API. These **must** match for collation to work. If the types are in different namespaces, then it's easy to retain the same name. If they aren't, then they need distinct names because it is enforced by the type system. In this case, API Versioning adds a second (but configurable) convention that supports the format `<Name>[V]#Controller`. As such, `OrderController` and `OrderV2Controller` will both have the logical name `"Order"` and be collated. There are some edge cases where you might not want that; for example `S3Controller`. 

`""` and `"V2"` are not the same so they will not be collated. You are also versioning by URL segment. By doing so, you have violated the _Uniform Interface_ REST constraint, which makes the API version part of the resource identifier. Since `1.0` is not collated into the set of API versions, the identifier cannot exist, therefore `404` is returned in the response. If you had `AddProblemDetails` enabled, you would probably see additional details indicating that the API version doesn't match/exist.

My recommendation for your next steps is to:

1. Use a more specific name; `Controller` is too generic
    a. If this is not possible due to backward compatibility, then apply `[ControllerName("controller")]` to `V2Controller`
2. Make sure your controller implementations correct and complete
3. If you're not using OpenAPI, don't both with `AddApiExplorer`; just use `AddMvc`

**Comment by mcastagna-sa** [comment]
Hi, thanks for taking your time to answer me. This is just an example i've put together just to show that is reproducible. I created the same issue in the grpc repo but i havent heard from them. (See: https://github.com/grpc/grpc-dotnet/issues/2500)
I can upload a zip with the code i posted above so you can check that commenting out that line shows the described behaviour. I appreciate your feedback though.

**Comment by mcastagna-sa** [comment]
I followed your suggestions and updated the project i wrote. I got the same behaviour. It would be great if we, at least, can rule out the versioning library as the problem.

[TestVersioningGRPC.zip](https://github.com/user-attachments/files/16948726/TestVersioningGRPC.zip)

**Comment by commonsensesoftware** [comment]
For a little clarity, does this problem exist for all gRPC endpoints? I still need to deep dive on this, but looking back at the history, the problem seems to be related to any gRPC endpoint. Also, does all of the routing work as expected? It appears that this may only be a problem for the API Explorer. I'll be using the repro to investigate further. Thanks.

**Comment by oleg-sviridov** [comment]
@commonsensesoftware I think this problem exists for all grpc endpoints. Regarding the routing I don't remember, sorry, it was too long ago and right now I don't have enough capacity to look again at the issue. Thank you

**Comment by commonsensesoftware** [comment]
I've done some additional investigation, and I think the issue is more nuanced than it appears. `.AddApiExplorer` is not directly the culprit. Since API Versioning `6.0`, you must also all `.AddMvc` (on `IApiVersioningBuilder`) to enable controller-based APIs or you only get Minimal APIs. `.AddApiExplorer` implicitly calls `.AddMvc` to ensure the required services are present. I was able to repro the issue by just calling `.AddMvc`.

Just calling `.AddMvc` alone will not cause the problem. Upon closer inspection, I see that you are logically overlapping the route templates, except that the controller has the `apiVersion` route constraint, whereas the gRPC template does not (and cannot). API version metadata and the route constraint are critical. Route matching does **not** happen with _magic strings_. As a result, `api/v2` will **not** match when endpoint selection occurs. By happenstance, it _almost_ could work.

Matching by URL segment is very tricky because there are no route candidates to select from. Internally, this process works by collating **all** of the route templates in a set of endpoints. _Any_ of the templates can then be used to extract the API version parameter from the request URL. It's yucky and slow, but it works. Since there is a `api/v{version:apiVersion}` for the controller definition, this template can be used to match the incoming API version. This is one of the many reasons I do **not** recommend versioning by URL segment; aside from it not being RESTful (it violates the _Uniform Interface_ constraint).

If the gRPC endpoint had the appropriate metadata **and** at least one other endpoint had a matching, grouped route template with the expected route constraint, then it _could_ match. The registration would need to look like:

```c#
app.MapGrpcService<GreeterService>().HasApiVersion( 1.0 );
```
>Interestingly, the required extensions appear to already be in place.

In theory, this would get you want you want and would match the gRPC endpoint or the controller based on API version. Sadly, after I got this far, I found other sharp edges. Ultimately, I'm going to chalk this up as a feature gap as opposed to a bug. This _is_ a scenario I would like to support, but I think it will be more involved to do right. I would prefer to punt this to the next version so I can start working on .NET 10; however, if you really need it in .NET 8, I'll consider backporting the feature to an `8.2` or something. It's been a while so I'm guessing this is not a burning issue.

I did notice that things work correctly if the gRPC service route does not overlap with a normal controller or Minimal API. If the routes can differ, that's one escape hatch. You could also _fake_ it by having the gRPC endpoint be on a different route and have a special _versioned_ controller route that always redirects to the target gRPC endpoint. It's definitely not ideal, but it's workable. It comes down to how you would like it to look for your clients.

I was able to get the gRPC endpoints to render and execute with OpenAPI and the Swagger UI as long as there are no routing issues. This is likely still not what you want. Since the gRPC endpoint is effectively _unversioned_ the API Explorer treats it the same way that it would a version-neutral API. This means that appears in every single document and API version. I thought I would get around it with `.WithGroupName("V1")`, but it didn't work. That could also be another edge case. It's uncommon to have truly _unversioned_ APIs and also show them in the Swagger UI. There _might_ be some way you can tweak it or otherwise get it to work.

Long story short, I think the _right_ thing to do is create a new tracking issue (I can do that) that outlines formally supporting gRPC which links back to this. I'm happy to keep the issue open until it releases.

**Comment by commonsensesoftware** [comment]
@oleg-sviridov I'm pretty sure the same thing will apply to `AddVersionedApiExplorer`. That is from the older, now deprecated, library. Are you still using that? I don't have any plans to backport that far. I'm already constrained as a one-man show and I don't really have the capacity to service more than _N-1_ LTS versions. It's been 3+ years now. I would strongly recommend upgrading/migrating as soon as you can. Happy to help facilitate the process through documentation or a [discussion](../discussions).

**Comment by commonsensesoftware** [comment]
@mcastagna-sa , @oleg-sviridov 

I'm trying to get a better sense of what people fully want out of this ask. Getting the routing to work is more straight forward than I thought, but the OpenAPI integration seems to be a bit unstable and stale in .NET 10. I found a bug in the gRPC integration that prevents API Versioning from working in an idiomatic way, but it can still work with things today. The following should work in .NET 8 or 10 without any additional changes.

The first _problem_ is that you're versioning by URL segment. This is not RESTful as it violates the _Uniform Interface_ constraint. This makes it much more difficult to support. I'm going to demonstrate how to make it work, but understand that other methods of versioning do not have this issue.

The first thing you need is create an `Interceptor`. This is required so you can wire up the URL segment to API Versioning. This is very similar, but not exactly the same, as how versioning by URL segment works for controllers (see [ApiVersionRouteConstraint](../blob/main/src/AspNetCore/WebApi/src/Asp.Versioning.Http/Routing/ApiVersionRouteConstraint.cs)). A key difference is that you cannot use traditional route constraints. It's not a hard requirement, but it logically makes sense to use the same parameter name for gRPC as you would for any other route template. This is configurable and retrievable via `ApiVersioningOptions.RouteConstraintName`. gRPC does not allow partial URL path segment matching and `v` is never part of the API version. The interceptor has to trim off the leading character and explicitly set the extracted value as the incoming `RawRequestedApiVersion`. If this matches a defined API Version, then the endpoint will be matched and routed to.

```c#
public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
    TRequest request,
    ServerCallContext context,
    UnaryServerMethod<TRequest, TResponse> continuation)
{
    var http = context.GetHttpContext();
    var options = http.RequestServices.GetRequiredService<IOptions<ApiVersioningOptions>>().Value;
    var name = options.RouteConstraintName;

    if (http.Request.RouteValues[name]?.ToString() is { } value)
    {
        var rawRequestedVersion = value.TrimStart('v', 'V');
        var feature = http.ApiVersioningFeature;
        var parser = http.RequestServices.GetRequiredService<IApiVersionParser>();

        feature.RouteParameter = name;
        feature.RawRequestedApiVersion = rawRequestedVersion;

        if (parser.TryParse(rawRequestedVersion, out var requestedVersion))
        {
            feature.RequestedApiVersion = requestedVersion;
        }
    }

    return await continuation(request, context);
}
```

You wire this up with:

```c#
builder.Services
        .AddGrpc(options => options.Interceptors.Add<ApiVersionInterceptor>())
        .AddJsonTranscoding();
```

You can now update your gRPC definition with:

```protobuf
service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {        
      get: "/api/{apiVersion}/fakecontroller/{name}/route"
    };
  }
}

message HelloRequest {
  string name = 1;
  optional string apiVersion = 2; // this is required for the interceptor even if it's not used
}
```

I found a bug in [GrpcServiceEndpointConventionBuilder](https://github.com/grpc/grpc-dotnet/blob/2e283d98f5861e31f8956be0c68aff9a4e29cb0a/src/Grpc.AspNetCore.Server/GrpcServiceEndpointConventionBuilder.cs#L24) which implements `IEndpointConventionBuilder` and `Apply`, but **not** `Finally`. The `Finally` method was added with a default interface implementation, which throws an exception by default 🤦🏽. API Versioning requires `Finally` in order to collate all of the defined API versions. The way to get around this is to perform the aggregation and apply the required metadata yourself.

```c#
var apiModel = new ApiVersionModel(new ApiVersion(1.0));
var endpointModel = new ApiVersionModel(new ApiVersion(1.0));
var metadata = new ApiVersionMetadata(apiModel, endpointModel, "Greeter");

app.MapGrpcService<GreeterService>().WithMetadata(metadata);
```

Now, when you request `/api/v1/fakecontroller/whatever/route`, the gRPC route will be correctly routed to.

If you add another version of your service and you version - say - by query string, you can do so with a lot less ceremony:

```protobuf
service Greeter2 {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {        
      get: "/api/fakecontroller/{name}/route"
    };
  }
}
```

```c#
public class GreeterService2(ILogger<GreeterService> logger) : Greeter2.Greeter2Base
{
    public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
    {
        var version = context.GetHttpContext().ApiVersioningFeature.RequestedApiVersion;
        logger.LogInformation("Saying hello to {Name}", request.Name);
        return Task.FromResult(new HelloReply { Message = $"Hello {request.Name} (v{version})" });
    }
}
```

```c#
var apiModel = new ApiVersionModel(new ApiVersion(3.0));
var endpointModel = new ApiVersionModel(new ApiVersion(3.0));
var metadata = new ApiVersionMetadata(apiModel, endpointModel, "Greeter");

app.MapGrpcService<GreeterService2>().WithMetadata(metadata);
```

You can now route to this new version of the service via `/api/fakecontroller/whatever/route?api-version=3.0`.

The overall setup can certainly be simplified, but the basic mechanics can work as they currently are. I need to post an issue (and maybe a PR) over to the gRPC repo to get `IEndpointConventionBuilder.Finally` supported. That would enable all of the metadata stuff to work as it is today. I'm not convinced I want to own an entire library just to bridge versioning by URL segment. This is achievable by a single class. I'm happy to provide an example or Gist instead, which was similarly how bridging Swashbuckle gaps were addressed.

I was struggling to get the current versions of the OpenAPI stuff to work properly with gRPC. If someone has a vanilla, unversioned gRPC with transcoding repro I can work backward from, I can investigate what - if anything - needs to change to support that integration.

**Comment by commonsensesoftware** [comment]
A small update on this issue. There's still some sharp edges, but [Grpc.AspNetCore.Server 2.80.0-pre1](https://www.nuget.org/packages/Grpc.AspNetCore.Server/2.80.0-pre1) contains the necessary fix to make the route extensions work. This means you can now simply have:

```c#
var app = build.Build();
var greeter = app.NewVersionedApi("Greeter");

greeter.MapGrpcService<GreeterService>().HasApiVersion(1.0);
greeter.MapGrpcService<GreeterService2>().HasApiVersion(3.0);
```

There's more to go for routing and potentially quite a bit more for any type of OpenAPI support to work, but there is some progress. It's still not 100% clear if the solution will require a new library of if things can be made to work with customization of how they are now.

Stay tuned.

**Comment by commonsensesoftware** [comment]
I've had a little bit of time and I was trying to get back to this. I was even excited to see some package updates with the hopes of some old blockers being removed. Unfortunately, the party came to halt when I discovered that Microsoft is dropping support for it. It is discussed at length in dotnet/aspnetcore#67134. The code was just officially removed in dotnet/aspnetcore#67919. Suffice to say, I don't think there is a path forward at this point.

I see there is some potential interest in forking that implementation and starting something anew. If that were to happen, I'm willing to revisit this issue. The advice I would give to any would be implementer is to **not** index solely on OpenAPI, Swagger, or Swashbuckle (though I doubt that will happen - again). Instead - love it or hate it, map to the API Explorer. This prevents bifurcation and opens a whole bunch of doors, including OpenAPI/Swagger support. This is what API Versioning does, not only because it predates OpenAPI/Swagger, but it opens other integration points that OpenAPI does not. I think of the API Explorer more like _Reflection for you web API_ and OpenAPI as how to represent the reflected types in a language-agnostic way.

The bridge between API Versioning and OpenAPI is quite small. Given some different choices or integration points with ASP.NET, there might not even need to be any direct bridges. I've avoided, and succeeded, at **not** having any direct dependencies on Swashbuckle, NSwag, Swagger, Scalar, etc for years now. API Explorer integration is likely the first, better long-term solution to API discovery and mapping to other formats such as OpenAPI.

cc: @adam8797, @atrokhovsergey

**Comment by adam8797** [comment]
@commonsensesoftware Ahh I see the deprecation got you too. I'm not as familiar with this issue, but I'm putting together https://github.com/adam8797/Grpc.Community.OpenApi with the remains of what they dropped. Its still pretty fresh, but I've at least got it building on its own. 

The original issue that got me looking at this package was https://github.com/dotnet/aspnetcore/issues/64879 so I'm not familiar with this issue in particular. 

If you've got thoughts, happy to discuss

**Comment by commonsensesoftware** [comment]
For completeness, I've double-checked and the OP setup does work with API Versioning and gRPC once the fix was merged on the gRPC side. There just isn't any API Explorer or OpenAPI/Swagger support.

Here's how I _think_ you'd want the final `*.proto` file. I'm highlight the version by URL segment and by query string methods, but you don't need to use both:

```proto
syntax = "proto3";

option csharp_namespace = "TestVersioningGRPC.Proto";
import "google/api/annotations.proto";

package greet;

service Greeter {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {
      get: "/api/{api_version}/fakecontroller/{name}/route"
    };
  }
}

message HelloRequest {
  string name = 1;

  // use 'json_name' for the 'api-version' query parameter name or register a different one
  string api_version = 2 [json_name = "api-version"];
}

message HelloReply {
  string message = 1;
}

service Greeter2 {
  rpc SayHello (HelloRequest) returns (HelloReply) {
    option (google.api.http) = {
      get: "/api/fakecontroller/{name}/route"
    };
  }
}
```

The interceptor was giving me false information because of `AssumeDefaultVersionWhenUnspecified = true` and using API version `1.0`. The interceptor will not run until **after** routing, which makes it worthless for the intent. I've come up with a way to make it work, but it's gross and will always be the slowest routing path. The URL segment method is already slower that other paths today. It considers all route templates collated by the routing system to extract the API version in the same way a constraint would (later on). This avoids hardcoding or knowing anything about what a user configured. In the case of gRPC, the route is defined outside of the routing system so that technique won't work. I'll have to nibble segment by segment on the request path to see if a match is found.  The only nice part is that this fallback approach will also work for API versions with a status that are used in segments, which currently don't work.

Assuming the versions are all separate classes, then implementations are now:

```c#
namespace TestVersioningGRPC.Services;

using Grpc.Core;
using TestVersioningGRPC.Proto;

// mapped using the url method
public class GreeterService(ILogger<GreeterService> logger) : Greeter.GreeterBase
{
    public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
    {
        logger.LogInformation("Saying hello to {Name}", request.Name);
        var version = context.GetHttpContext().ApiVersioningFeature.RawRequestedApiVersion; // ← won't have 'v' prefix
        return Task.FromResult(new HelloReply { Message = $"Hello {request.Name} (v{version})" });
    }
}

// mapped using the query string method
public class GreeterService2(ILogger<GreeterService> logger) : Greeter2.Greeter2Base
{
    public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
    {
        logger.LogInformation("Saying hello to {Name}", request.Name);
        return Task.FromResult(new HelloReply { Message = $"Hello {request.Name} (v{request.ApiVersion})" });
    }
}
```

```c#
namespace TestVersioningGRPC.Controllers;

using Asp.Versioning;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[ApiVersion(2.0)]
[Produces("application/json")]
[Route("api/v{version:apiVersion}/fakecontroller")]
public class FakeV2Controller : Controller
{
    [HttpGet("{param}/route")]
    public IActionResult GetEndpoint(string param, ApiVersion apiVersion) => Ok($"hello {param} (v{apiVersion})");
}
```

Finally, you put it all together like this:

```c#
var builder = WebApplication.CreateBuilder(args);
var services = builder.Services;

services.AddApiVersioning().AddMvc();
services.AddGrpc().AddJsonTranscoding();

var app = builder.Build();
var greeter = app.NewVersionedApi("Greeter");

greeter.MapGrpcService<GreeterService>().HasApiVersion(1.0);
greeter.MapGrpcService<GreeterService2>().HasApiVersion(3.0);

app.Run();
```

The results will be:

- `api/v1/fakecontroller/test/route`  → `{"message":"Hello test (v1)"}`
- `api/v2/fakecontroller/test/route`  → `"hello test (v2)"`
- `api/fakecontroller/test/route?api-version=3.0`  → `{"message":"Hello test (v3.0)"}`


In conclusion, you can make the query string today without any changes. The header and media type methods don't make sense for gRPC. If you want to use the URL segment method, you'll have to wait for a patch, which I hope to have out this week. That will at least give you versioned gRPC. 😉

**Comment by commonsensesoftware** [comment]
@adam8797, this issue is a combination of versioning gRPC **and** providing OpenAPI support. This project is rooted in the API Explorer, which most other libraries use as the foundational piece for OpenAPI. Where there's a gap, they are filled in here, by the community, or in your code. IMHO, if there were API Explorer like support for gRPC endpoints, then most of it would just work well together as is. You might not even need to do any OpenAPI/Swagger stuff directly. Food for thought. It's a different approach from what Microsoft did. They went straight for the gRPC-specific solution. If you go straight to OpenAPI, then it _might_ not work with API versioning. Now, things will be back in a place where people want versioned gRPC endpoints **and** OpenAPI, but there is little-to-no common ground. For historical prior art, that is more or less the situation with OData who provides _some_ OpenAPI/Swagger support, but it doesn't support versioning. The versioned OData support here is vastly more feature rich.

**Comment by commonsensesoftware** [comment]
`10.2.0` has dropped and there is now end-to-end support for API Versioning with gRPC, which includes the API Explorer and OpenAPI. The shiny new wiki will take you through [API Versioning with gRPC](https://dotnet.github.io/aspnet-api-versioning/aspnet-core/grpc/overview.html). You can find a complete working example in the [gRPC OpenAPI Example](../tree/main/examples/AspNetCore/WebApi/GrpcOpenApiExample). The example also illustrates interleaving with controllers too.

The new packages are currently in `preview.1` because I have no idea what things I may have missed, but this should be a very strong foundation. I didn't notice any gaps, but I need your help to let me know otherwise. If you do find issues, please open a new specific issue related to the issue. I'm about inclined to close this particular issue because I think the OP has been addressed.

**Comment by commonsensesoftware** [comment]
I have officially promoted the gRPC support to release. Since there has been any reported issues or additional feedback in the past month, I believe this issue has reached its conclusion and is ready to close. AFAIK, the OP has been addressed and more. If any new issues arise, please open a new issue.

We're about 2 months away from .NET 11. If you discover any missing or incomplete features, I'll be happy to consider them before the final release.
